src/libstore/ssh.cc

```diff
@@ -86,6 +92,18 @@ std::unique_ptr<SSHMaster::Connection> SSHMaster::startCommand(const std::string
     in.readSide = -1;
     out.writeSide = -1;
 
+    // Wait for the SSH connection to be established,
+    // So that we don't overwrite the password prompt with our progress bar.
+    if (!fakeSSH) {
+        std::string reply;
+        try {
+            reply = readLine(out.readSide.get());
+        } catch (EndOfFile & e) { }
+
+        if (reply != "started")
+            throw Error("failed to start SSH connection to '%s'", host);
+    }
```

Comment:
Maybe it's just the git diff misleading me, but if seems that this runs for each ssh call. However, the `LocalCommand` setting is ignored if `ssh` is reusing an existing master connection (which is generally the case in Nix), so maybe this should all stay in `startMaster()` as it would get stuck otherwise

Reply:
Good catch!